### ISSUE TITLE
src/mux_h2.c: uninitialized variable suspected by coverity

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#2014]
### Tool Name and Version

coverity

### Code Report

```plain
long report 


5077static size_t h2s_snd_fhdrs(struct h2s *h2s, struct htx *htx)
5078{
    1. var_decl: Declaring variable list without initializer.
5079        struct http_hdr list[global.tune.max_http_hdr];
5080        struct h2c *h2c = h2s->h2c;
5081        struct htx_blk *blk;
5082        struct buffer outbuf;
5083        struct buffer *mbuf;
5084        struct htx_sl *sl;
5085        enum htx_blk_type type;
5086        int es_now = 0;
5087        int ret = 0;
5088        int hdr;
5089
    2. Condition trace_h2.state != TRACE_STATE_STOPPED, taking false branch.
5090        TRACE_ENTER(H2_EV_TX_FRAME|H2_EV_TX_HDR, h2c->conn, h2s);
5091
5092        /* get the start line (we do have one) and the rest of the headers,
5093         * that we dump starting at header 0 */
5094        sl = NULL;
5095        hdr = 0;
    3. Condition blk, taking true branch.
    6. Condition blk, taking true branch.
    9. Condition blk, taking true branch.
    23. Condition blk, taking true branch.
    26. Condition blk, taking false branch.
5096        for (blk = htx_get_head_blk(htx); blk; blk = htx_get_next_blk(htx, blk)) {
5097                type = htx_get_blk_type(blk);
5098
    4. Condition type == HTX_BLK_UNUSED, taking true branch.
    7. Condition type == HTX_BLK_UNUSED, taking true branch.
    10. Condition type == HTX_BLK_UNUSED, taking false branch.
    24. Condition type == HTX_BLK_UNUSED, taking true branch.
5099                if (type == HTX_BLK_UNUSED)
    5. Continuing loop.
    8. Continuing loop.
    25. Continuing loop.
5100                        continue;
5101
    11. Condition type == HTX_BLK_EOH, taking false branch.
5102                if (type == HTX_BLK_EOH)
5103                        break;
5104
    12. Condition type == HTX_BLK_HDR, taking false branch.
5105                if (type == HTX_BLK_HDR) {
5106                        BUG_ON(!sl); /* The start-line mut be defined before any headers */
5107                        if (unlikely(hdr >= sizeof(list)/sizeof(list[0]) - 1)) {
5108                                TRACE_ERROR("too many headers", H2_EV_TX_FRAME|H2_EV_TX_HDR|H2_EV_H2S_ERR, h2c->conn, h2s);
5109                                goto fail;
5110                        }
5111
5112                        list[hdr].n = htx_get_blk_name(htx, blk);
5113                        list[hdr].v = htx_get_blk_value(htx, blk);
5114                        hdr++;
5115                }
    13. Condition type == HTX_BLK_RES_SL, taking true branch.
5116                else if (type == HTX_BLK_RES_SL) {
    14. Condition sl != NULL, taking false branch.
5117                        BUG_ON(sl); /* Only one start-line expected */
5118                        sl = htx_get_blk_ptr(htx, blk);
5119                        h2s->status = sl->info.res.status;
    15. Condition h2s->status == 204, taking true branch.
5120                        if (h2s->status == 204 || h2s->status == 304)
5121                                h2s->flags |= H2_SF_BODYLESS_RESP;
    16. Condition h2s->status < 100, taking false branch.
    17. Condition h2s->status > 999, taking false branch.
5122                        if (h2s->status < 100 || h2s->status > 999) {
5123                                TRACE_ERROR("will not encode an invalid status code", H2_EV_TX_FRAME|H2_EV_TX_HDR|H2_EV_H2S_ERR, h2c->conn, h2s);
5124                                goto fail;
5125                        }
    18. Condition h2s->status == 101, taking false branch.
5126                        else if (h2s->status == 101) {
5127                                if (unlikely(h2s->flags & H2_SF_EXT_CONNECT_RCVD)) {
5128                                        /* If an Extended CONNECT has been received, we need to convert 101 to 200 */
5129                                        h2s->status = 200;
5130                                        h2s->flags &= ~H2_SF_EXT_CONNECT_RCVD;
5131                                }
5132                                else {
5133                                        /* Otherwise, 101 responses are not supported in H2, so return a error (RFC7540#8.1.1) */
5134                                        TRACE_ERROR("will not encode an invalid status code", H2_EV_TX_FRAME|H2_EV_TX_HDR|H2_EV_H2S_ERR, h2c->conn, h2s);
5135                                        goto fail;
5136                                }
5137                        }
    19. Condition h2s->flags & 1024, taking true branch.
    20. Condition h2s->status >= 300, taking false branch.
5138                        else if ((h2s->flags & H2_SF_BODY_TUNNEL) && h2s->status >= 300) {
5139                                /* Abort the tunnel attempt */
5140                                h2s->flags &= ~H2_SF_BODY_TUNNEL;
5141                                h2s->flags |= H2_SF_TUNNEL_ABRT;
5142                        }
    21. Falling through to end of if statement.
5143                }
5144                else {
5145                        TRACE_ERROR("will not encode unexpected htx block", H2_EV_TX_FRAME|H2_EV_TX_HDR|H2_EV_H2S_ERR, h2c->conn, h2s);
5146                        goto fail;
5147                }
    22. Jumping back to the beginning of the loop.
5148        }
5149
5150        /* The start-line me be defined */
    27. Condition !sl, taking false branch.
5151        BUG_ON(!sl);
5152
5153        /* marker for end of headers */
5154        list[hdr].n = ist("");
5155
5156        mbuf = br_tail(h2c->mbuf);
5157 retry:
    28. Condition !h2_get_buf(h2c, mbuf), taking false branch.
5158        if (!h2_get_buf(h2c, mbuf)) {
5159                h2c->flags |= H2_CF_MUX_MALLOC;
5160                h2s->flags |= H2_SF_BLK_MROOM;
5161                TRACE_STATE("waiting for room in output buffer", H2_EV_TX_FRAME|H2_EV_TX_HDR|H2_EV_H2S_BLK, h2c->conn, h2s);
5162                return 0;
5163        }
5164
5165        chunk_reset(&outbuf);
5166
    29. Condition 1, taking true branch.
    52. Condition 1, taking true branch.
    80. Condition 1, taking true branch.
5167        while (1) {
5168                outbuf = b_make(b_tail(mbuf), b_contig_space(mbuf), 0, 0);
    30. Condition outbuf.size >= 9, taking true branch.
    53. Condition outbuf.size >= 9, taking true branch.
    81. Condition outbuf.size >= 9, taking true branch.
5169                if (outbuf.size >= 9 || !b_space_wraps(mbuf))
    31. Breaking from loop.
    54. Breaking from loop.
    82. Breaking from loop.
5170                        break;
5171        realign_again:
5172                b_slow_realign(mbuf, trash.area, b_data(mbuf));
    51. Jumping back to the beginning of the loop.
    79. Jumping back to the beginning of the loop.
5173        }
5174
    32. Condition outbuf.size < 9, taking false branch.
    55. Condition outbuf.size < 9, taking false branch.
    83. Condition outbuf.size < 9, taking false branch.
5175        if (outbuf.size < 9)
5176                goto full;
5177
5178        /* len: 0x000000 (fill later), type: 1(HEADERS), flags: ENDH=4 */
5179        memcpy(outbuf.area, "\x00\x00\x00\x01\x04", 5);
5180        write_n32(outbuf.area + 5, h2s->id); // 4 bytes
5181        outbuf.data = 9;
5182
    33. Condition (h2c->flags & (6291456U /* 0x200000 | 0x400000 */)) == 2097152, taking true branch.
    56. Condition (h2c->flags & (6291456U /* 0x200000 | 0x400000 */)) == 2097152, taking true branch.
    84. Condition (h2c->flags & (6291456U /* 0x200000 | 0x400000 */)) == 2097152, taking true branch.
5183        if ((h2c->flags & (H2_CF_SHTS_UPDATED|H2_CF_DTSU_EMITTED)) == H2_CF_SHTS_UPDATED) {
5184                /* SETTINGS_HEADER_TABLE_SIZE changed, we must send an HPACK
5185                 * dynamic table size update so that some clients are not
5186                 * confused. In practice we only need to send the DTSU when the
5187                 * advertised size is lower than the current one, and since we
5188                 * don't use it and don't care about the default 4096 bytes,
5189                 * we only ack it with a zero size thus we at most have to deal
5190                 * with this once. See RFC7541#4.2 and #6.3 for the spec, and
5191                 * below for the whole context and interoperability risks:
5192                 * https://lists.w3.org/Archives/Public/ietf-http-wg/2021OctDec/0235.html
5193                 */
    34. Condition b_room(&outbuf) < 1, taking false branch.
    57. Condition b_room(&outbuf) < 1, taking false branch.
    85. Condition b_room(&outbuf) < 1, taking false branch.
5194                if (b_room(&outbuf) < 1)
5195                        goto full;
5196                outbuf.area[outbuf.data++] = 0x20; // HPACK DTSU 0 bytes
5197
5198                /* let's not update the flags now but only once the buffer is
5199                 * really committed.
5200                 */
5201        }
5202
5203        /* encode status, which necessarily is the first one */
    35. Condition !hpack_encode_int_status(&outbuf, h2s->status), taking false branch.
    58. Condition !hpack_encode_int_status(&outbuf, h2s->status), taking false branch.
    86. Condition !hpack_encode_int_status(&outbuf, h2s->status), taking false branch.
5204        if (!hpack_encode_int_status(&outbuf, h2s->status)) {
5205                if (b_space_wraps(mbuf))
5206                        goto realign_again;
5207                goto full;
5208        }
5209
    36. Condition trace_h2.verbosity >= 4, taking false branch.
    59. Condition trace_h2.verbosity >= 4, taking false branch.
    87. Condition trace_h2.verbosity >= 4, taking false branch.
5210        if ((TRACE_SOURCE)->verbosity >= H2_VERB_ADVANCED) {
5211                char sts[4];
5212
5213                h2_trace_header(ist(":status"), ist(ultoa_r(h2s->status, sts, sizeof(sts))),
5214                                    H2_EV_TX_FRAME|H2_EV_TX_HDR, ist(TRC_LOC), __FUNCTION__,
5215                                    h2c, h2s);
5216        }
5217
5218        /* encode all headers, stop at empty name */
    37. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
    40. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
    60. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
    63. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
    66. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
    88. Condition hdr < sizeof (list) / 32UL /* sizeof (list[0]) */, taking true branch.
5219        for (hdr = 0; hdr < sizeof(list)/sizeof(list[0]); hdr++) {
5220                /* these ones do not exist in H2 and must be dropped. */
    38. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking true branch.
    41. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    42. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"proxy-connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    43. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"keep-alive" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    44. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"upgrade" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    45. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"transfer-encoding" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    61. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking true branch.
    64. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking true branch.
    67. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    68. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"proxy-connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    69. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"keep-alive" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    70. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"upgrade" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    71. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"transfer-encoding" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    89. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    90. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"proxy-connection" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    91. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"keep-alive" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    92. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"upgrade" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    93. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"transfer-encoding" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
5221                if (isteq(list[hdr].n, ist("connection")) ||
5222                    isteq(list[hdr].n, ist("proxy-connection")) ||
5223                    isteq(list[hdr].n, ist("keep-alive")) ||
5224                    isteq(list[hdr].n, ist("upgrade")) ||
5225                    isteq(list[hdr].n, ist("transfer-encoding")))
    39. Continuing loop.
    62. Continuing loop.
    65. Continuing loop.
5226                        continue;
5227
5228                /* Skip all pseudo-headers */
    46. Condition *list[hdr].n.ptr == ':', taking false branch.
    72. Condition *list[hdr].n.ptr == ':', taking false branch.
    94. Condition *list[hdr].n.ptr == ':', taking false branch.
5229                if (*(list[hdr].n.ptr) == ':')
5230                        continue;
5231
    47. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
    73. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking true branch.
    95. Condition isteq(list[hdr].n, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});})), taking false branch.
5232                if (isteq(list[hdr].n, ist("")))
    74. Breaking from loop.
5233                        break; // end
5234
    48. Condition !h2_encode_header(&outbuf, list[hdr].n, list[hdr].v, 71303168UL /* (1ULL << 22) | (1ULL << 26) */, ({...; struct ist({.ptr = __x, .len = 1 ? (0 /* (void *)"src/mux_h2.c:5236" == NULL */) ? 0 : __builtin_strlen(__x) : ({...; __l;})});}), <anonymous>, h2c, h2s), taking true branch.
    96. uninit_use_in_call: Using uninitialized value list[hdr].v.len when calling h2_encode_header. [show details]
    CID 1504179 (#1-2 of 2): Uninitialized pointer read (UNINIT)97. uninit_use_in_call: Using uninitialized value list[hdr].v.ptr when calling h2_encode_header. [show details]
5235                if (!h2_encode_header(&outbuf, list[hdr].n, list[hdr].v, H2_EV_TX_FRAME|H2_EV_TX_HDR,
5236                                      ist(TRC_LOC), __FUNCTION__, h2c, h2s)) {
5237                        /* output full */
    
```
```


### Additional Information

_No response_

### Output of `haproxy -vv`

```plain
no
```

**Comment by capflam** [comment]
It is also a duplicate of #1084. The `if` statement at line 5232 breaks the loop when `list[hdr].n` is an empty string. It is the only case to have `list[hdr].v` uninitialized. Thus, a line 5235, when `h2_encode_header()` is called, `.n` and `.v` are both valid.
